### PR TITLE
Propose specification on Multi User User-Agent Extension

### PR DESCRIPTION
**Pull request by der-lyse**
As promised, here's the proposal for the User-Agent extension in multi user clients.

### CODE REVIEW

File: docs/_posts/2021-01-06-useragentextension.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
+---
+layout: page
+title: "Multi User User-Agent Extension"
```
Comment (der-lyse) [inline docs/_posts/2021-01-06-useragentextension.md:3]: Does anyone have a better title?

File: docs/_posts/2021-01-06-useragentextension.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+## Format
+
+Depending on the number of followers on a multi user instance there are three
+different formats to be used in the `User-Agent` HTTP request header.
+
+### Single Follower
+
+If there's only a single follower, the original twtxt specification on
+[Discoverability](https://twtxt.readthedocs.io/en/latest/user/discoverability.html)
+should be followed, to be backwards-compatible:
```
Comment (der-lyse) [inline docs/_posts/2021-01-06-useragentextension.md:32]: "should" or "must"?

File: docs/_posts/2021-01-06-useragentextension.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+Nicks should be in alphabetical order.
+
+The support information is optional and should point to a page were the client
+owner can be contacted.
+
+### Six Or More Followers
+
+To avoid `User-Agent` headers getting too large, there should be a limit on the
+number of nicks to be included. The exact number when to switch formats is up to
+the client author or operator. When six or more users follow the same feed, the
```
Comment (der-lyse) [inline docs/_posts/2021-01-06-useragentextension.md:78]: Do you agree on letting others decide on the their favorite number or do we want to define it here exactly?

File: docs/_posts/2021-01-06-useragentextension.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+For example:
+
+```
+twtxt/0.1.0@cdd6014 (Pod: example.com Followers: somebody someoneelse user3 user4 user5 and 3 more... https://example.com/whoFollows?uri=https://example.com/twtxt.txt&nick=joe&token=R9nWDD23u Support: https://example.com/support)
+```
+
+### Who Follows Resource
+
+When requested with the `Accept: application/json` header, this resource should
```
Comment (der-lyse) [inline docs/_posts/2021-01-06-useragentextension.md:96]: "should" or "must"?

File: docs/_posts/2021-01-06-useragentextension.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+provide a JSON object with nicks as keys and their *twtxt.txt* file URLs as
+values. The mapping should contain all followers, including those who are
```
Comment (der-lyse) [inline docs/_posts/2021-01-06-useragentextension.md:98]: "should" or "must"?

File: docs/_posts/2021-01-06-useragentextension.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+  "user5": "https://example.com/user/user5/twtxt.txt",
+  "user6": "https://example.com/user/user6/twtxt.txt",
+  "user7": "https://example.com/user/user7/twtxt.txt",
+  "user8": "https://example.com/user/user8/twtxt.txt"
+}
+```
+
+## Security Considerations
+
+Users of multi user clients should have the option to keep their following list
```
Comment (der-lyse) [inline docs/_posts/2021-01-06-useragentextension.md:123]: "should" or "must"?

File: docs/_posts/2021-01-06-useragentextension.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+secret and thus to hide themselves from both the `User-Agent` as well as Who
+Follows Resource.
+
+The Who Follows Resource could be easily guessable and thus should be somehow
```
Comment (der-lyse) [inline docs/_posts/2021-01-06-useragentextension.md:127]: "should" or "must"?

File: docs/_posts/2021-01-06-useragentextension.md [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+protected to not publicly disclose the followers of a certain feed to
+unauthorized third parties. Keep in mind, the `User-Agent` header is only
+available to the feed owner or web server operator. It should not be possible
```
Comment (der-lyse) [inline docs/_posts/2021-01-06-useragentextension.md:130]: "should" or "must"?

File: docs/_posts/2021-01-06-useragentextension.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+twtxt/<version> (Pod: <hostname> Followers: <nick1> <nick2> <nick3> <nick4> <nick5> and <number> more... <url> Support: <url>)
+```
+
+The `<number>` specifies the amount of users, which are excluded from the nick
+list, but which can be obtained from the given URL.
+
+For example:
+
+```
+twtxt/0.1.0@cdd6014 (Pod: example.com Followers: somebody someoneelse user3 user4 user5 and 3 more... https://example.com/whoFollows?uri=https://example.com/twtxt.txt&nick=joe&token=R9nWDD23u Support: https://example.com/support)
```
Comment (der-lyse) [inline docs/_posts/2021-01-06-useragentextension.md:91]: This URL is quite long and doesn't save any space so far with low follower numbers. Maybe the token should be internally mapped to the feed URL. I also believe the nick is not necessary here, but I haven't confirmed that in the code.

File: docs/_posts/2021-01-06-useragentextension.md (hunk 1 above)
Comment (prologic) [reply docs/_posts/2021-01-06-useragentextension.md:3]: Client Identifier?

File: docs/_posts/2021-01-06-useragentextension.md (hunk 2 above)
Comment (prologic) [reply docs/_posts/2021-01-06-useragentextension.md:32]: I would say _should_; to the best of my knowledge, the Twtxt backend that powers twtxt.net, twt.u53.net, txt.sour.is and others are the only clients that _actually_ automatically handle the detection of `User-Agent`(s) or Client Identifiers.

File: docs/_posts/2021-01-06-useragentextension.md (hunk 3 above)
Comment (prologic) [reply docs/_posts/2021-01-06-useragentextension.md:78]: I _think_ this could be left to interpretation. I don't think it matters a whole lot since in the end it's just a comma separated list of usernames from that pod.

File: docs/_posts/2021-01-06-useragentextension.md (hunk 4 above)
Comment (prologic) [reply docs/_posts/2021-01-06-useragentextension.md:96]: `must` -- because content-negotiation?

File: docs/_posts/2021-01-06-useragentextension.md (hunk 5 above)
Comment (prologic) [reply docs/_posts/2021-01-06-useragentextension.md:98]: _should_

File: docs/_posts/2021-01-06-useragentextension.md (hunk 6 above)
Comment (prologic) [reply docs/_posts/2021-01-06-useragentextension.md:123]: _should_ -- But I'm not 100% sure we implement this correctly?

File: docs/_posts/2021-01-06-useragentextension.md (hunk 7 above)
Comment (prologic) [reply docs/_posts/2021-01-06-useragentextension.md:127]: hmmm not sure on this one, I sort of learn towards "must". twtxt.net implements a short-lived token here. Not sure? It's not exactly that secure though.

File: docs/_posts/2021-01-06-useragentextension.md (hunk 8 above)
Comment (prologic) [reply docs/_posts/2021-01-06-useragentextension.md:130]: probably "shoud"

File: docs/_posts/2021-01-06-useragentextension.md (hunk 9 above)
Comment (prologic) [reply docs/_posts/2021-01-06-useragentextension.md:91]: Yes I _think_ you're right. This could just be a token that maps to that feed. That's all we need really.

File: docs/_posts/2021-01-06-useragentextension.md (hunk 1 above)
Comment (der-lyse) [reply docs/_posts/2021-01-06-useragentextension.md:3]: "Multi User Client Identifier User-Agent Extension"?

File: docs/_posts/2021-01-06-useragentextension.md (hunk 6 above)
Comment (der-lyse) [reply docs/_posts/2021-01-06-useragentextension.md:123]: What I've seen so far in the code, it looks correct to me.

File: docs/_posts/2021-01-06-useragentextension.md (hunk 7 above)
Comment (der-lyse) [reply docs/_posts/2021-01-06-useragentextension.md:127]: I agree with "must". Obviously, the longer the token itself and the shorter its livetime, the more secure. But I don't think we need to go totally crazy on this. I certainly wouldn't go below ten minutes of validity when polling every five minutes, to still be able to manually visit the URL. Another countermeasure would be to drop all the tokens once there are three wrong attempts to guess them. However, then the legitimate user could be locked out easily forever by malicious third parties.

File: docs/_posts/2021-01-06-useragentextension.md [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+different formats to be used in the `User-Agent` HTTP request header.
+
+### Single Follower
+
+If there's only a single follower, the original twtxt specification on
+[Discoverability](https://twtxt.readthedocs.io/en/latest/user/discoverability.html)
+should be followed, to be backwards-compatible:
+
+```
+twtxt/<version> (+<source.url>; @<source.nick>)
```
Comment (der-lyse) [inline docs/_posts/2021-01-06-useragentextension.md:35]: I noticed that according to the original spec the client software should always be "twtxt" and nothing else (or am I misreading it?), which twtxt.net conforms to, however, several other clients do not. Should we relax this to `<client>` to also "promote" other clients?

File: docs/_posts/2021-01-06-useragentextension.md (hunk 5 above)
Comment (vain) [reply docs/_posts/2021-01-06-useragentextension.md:98]: (Not sure I understand: What’s the consequence of “should” here? Would it be legal for the server to omit some followers? Why would that be needed?)

File: docs/_posts/2021-01-06-useragentextension.md (hunk 10 above)
Comment (vain) [reply docs/_posts/2021-01-06-useragentextension.md:35]: I get the impression that `twtxt` is not meant as a fixed string for *all* clients here. [The original spec](https://twtxt.readthedocs.io/en/stable/user/discoverability.html) ends with: “Other clients are encouraged to use the same format.” So, I take it that `twtxt` in this case refers to “twtxt, the reference implementation”.

So, yeah, I guess this should read something like `<client.name>/<version> (+<source.url>; @<source.nick>)`?

File: docs/_posts/2021-01-06-useragentextension.md (hunk 5 above)
Comment (der-lyse) [reply docs/_posts/2021-01-06-useragentextension.md:98]: As far as I understand it, yes, it would be legal to omit some followers. Other than the two obvious cases, I can't think of a real use case to only serve a partial followers list:

1. Users don't want to disclose their identities as explained in the "Security Considerations" section down below.
2. User's already being listed in the UA header could be missing in this JSON object (although this is kind of silly).

So maybe "must" is the way to go.

File: docs/_posts/2021-01-06-useragentextension.md (hunk 10 above)
Comment (der-lyse) [reply docs/_posts/2021-01-06-useragentextension.md:35]: Very nice interpretation, I like it!

Comment (prologic) [reply docs/_posts/2021-01-06-useragentextension.md:35]: Yeah I agree with @vain here. I don't think all clients need to be called "twtxt/" 🤣

**Comment by prologic** [comment]
For simplicity I'm actually thinking it _might_ be better if we just support two `User-Agent` formats:

- Single-User follower as per the original spec
- Multi-User / Multi-followers with a token-based callback.

And drop the "in-between".

Thoughts?

File: docs/_posts/2021-01-06-useragentextension.md (hunk 8 above)
Comment (der-lyse) [reply docs/_posts/2021-01-06-useragentextension.md:130]: After sleeping on this I believe we should go with "must not" rather than "should not", to force the client author to do some basic work and help the data protection.

**Comment by der-lyse** [comment]
That would certainly help the simplicity aspect, that's for sure. So I quite like that. We should also condense the format to something much more compact, such as:

```
<client.name>/<client.version> (~<who-follows.url>)
```

Here, the idea is to indicate the Who Follows Resource with some other character, e.g. `~` (or whatever, I'm no expert on UAs, so maybe there's a better choice) to make it easily distinguishable from the single user client, e.g.

```
twtxt/0.1.0@abcdefg (~https://example.com/whoFollows?token=randomtoken123)
```

Not sure whether we actually need the support URL, what's the reason behind that? It's certainly nice, but in which cases would one reach out to the client operator?

**Comment by der-lyse** [comment]
I incorporated some of your feedback, but I haven't touched the "in-between" UA format and I'm not sure about the title yet. Maybe we'll find a little bit more compact one.

**Comment by prologic** [comment]
> Not sure whether we actually need the support URL, what's the reason behind that? It's certainly nice, but in which cases would one reach out to the client operator?

It was added as a "work-around" for lack of a "who follows" callback URL. But also just in case you need to contact the pod owner for some reason?

**Comment by der-lyse** [comment]
> It was added as a "work-around" for lack of a "who follows" callback URL.

Ah, I see.

> But also just in case you need to contact the pod owner for some reason?

Nice bots certainly do that, but I'm not entirely sure this is super useful in our case as well. Please don't get me wrong, I don't say, it's worthless, I just don't know. I'm happy keeping it, I just want to learn about situations where it can come in handy. Maybe sys admins deeply watching their access logs can chime in to provide some use cases.

The only thing I can think of right now is to ask the client operator to reduce the sync interval for a set of feeds, e.g. because of expensive traffic or low bandwidth locations.

**Comment by prologic** [comment]
Shall we merge this soon or are we waiting on me to make the aforementioned changes I promised? If so, I'll file an issue and try to find some time to simplify the "who follows" implementation.

**Comment by der-lyse** [comment]
If you're gonna make the proposed "just two UA forms" change, I reckon it would be ideal to adjust the spec for this scenario now, what do you think?

**Comment by prologic** [comment]
I agree

**Comment by der-lyse** [comment]
Before going with the [changes proposed in a comment](https://github.com/jointwt/twtxt/pull/326#issuecomment-756175697), I checked both the [German](https://de.wikipedia.org/wiki/User_Agent) and [English User-Agent Wikipedia articles](https://en.wikipedia.org/wiki/User_agent), the appropriate sections on [header fields in general](https://tools.ietf.org/html/rfc7230#section-3.2) as well as the [User-Agent header specifically of the HTTP spec](https://tools.ietf.org/html/rfc7231#section-5.5.3), but I didn't find anything useful to our situation. From all my other research only a single piece turned up to be interesting: According to [this guy on StackExchange](https://webmasters.stackexchange.com/a/51891), most web spiders do not prefix their support contact with a plus sign (`+`). At least back in 2013.

It seems to me, that there are no real rules out there to follow. Everybody is doing their own things.

So I see nothing wrong with the `+` and `~` prefixes in general. The only bad thing is, that the `+` in the Single User format changes to a `~` in the Multi User format for the *interesting* URL. However, they're pointing to different things, so the inconsistency is probably not too terrible. I'm totally open to other prefix ideas.

What do you reckon @prologic, @vain, @JonLundy or anybody else?

File: docs/_posts/2021-01-06-useragentextension.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+twtxt/1.2.3 (+https://example.com/twtxt.txt; @somebody)
+```
+
+### Multiple Followers
+
+Starting with a second follower, the format changes. It aims to be fairly
+compact:
+
+```
+<client.name>/<client.version> (~<who-follows.url>; +<support.url>)
```
Comment (JonLundy) [inline docs/_posts/2021-01-06-useragentextension.md:50]: rather then conflate `+<follower-url>` and `+<support-url>`, have a `contact=email@example.com`

**Comment by vain** [comment]
A dedicated `X-` header would be the cleanest solution.

*But* it would require you to run server-side code to evaluate this header. Or at least a special web server config. I think OpenBSD httpd is not able to log custom headers.

The “beauty” of that user agent thingy is that it’s super easy to work with. The web server logs it anyway and I can just do a fancy grep.

Me, personally, I’m fine with the current draft. :)

**Comment by JonLundy** [comment]
@vain i agree. other than probably the suggestion to not reuse the + for support url.

File: docs/_posts/2021-01-06-useragentextension.md (hunk 11 above)
Comment (der-lyse) [reply docs/_posts/2021-01-06-useragentextension.md:50]: I like that a lot.

**Comment by der-lyse** [comment]
[RFC 6648](https://tools.ietf.org/html/rfc6648) deprecates the new introduction of `X-` prefixed custom headers, so we could simply use `Twt-Follows`/`Twt-Follow-Url` instead. However, I absolutely agree with the drawbacks of this approach as pointed out by @vain. Mixing @JonLundy's `contact=` into

```
User-Agent: twtxt/0.1.0 (~https://example.com/whoFollows?token=randomtoken123; contact=support@example.com)
```

is probably the way to go.

File: docs/_posts/2021-01-06-useragentextension.md (hunk 11 above)
Comment (prologic) [reply docs/_posts/2021-01-06-useragentextension.md:50]: Hmmm we don't actually expose the Pod Owner/Operator's email address though. Yes it is stored in configuration, but not ever shown anywhere. Its used by the /support form and /abuse reporting. I'm not sure I'd want to do this for spam reasons :)

**Review by prologic — APPROVED** [review]

**Comment by JonLundy** [comment]
@prologic while i agree that emails being an issue for spam, i is a common convention for reporting abuse like DKIM/SPF does with email.. though i suppose a contact web page could be an alternative like HTTP security headers use for enforcement reporting.

**Comment by prologic** [comment]
> @prologic while i agree that emails being an issue for spam, i is a common convention for reporting abuse like DKIM/SPF does with email.. though i suppose a contact web page could be an alternative like HTTP security headers use for enforcement reporting.

So we're okay with just display the pod owner's email here?

**Comment by prologic** [comment]
> A dedicated `X-` header would be the cleanest solution.
> 
> _But_ it would require you to run server-side code to evaluate this header. Or at least a special web server config. I think OpenBSD httpd is not able to log custom headers.
> 
> The “beauty” of that user agent thingy is that it’s super easy to work with. The web server logs it anyway and I can just do a fancy grep.
> 
> Me, personally, I’m fine with the current draft. :)

Yeah I tend to think we should stick to the UA here for compatibility and accessibility. Abuse a HTTP header here will likely just go unnoticed 😀

**Comment by der-lyse** [comment]
I changed the spec to explicitly accept both e-mail addresses as well as support URLs as valid client contact information.

File: docs/_posts/2021-01-06-useragentextension.md [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+compact:
+
+```
+<client.name>/<client.version> (~<who-follows.url>; contact=<client.contact>)
+```
+
+For example:
+
+```
+twtxt/0.1.0@abcdefg (~https://example.com/whoFollows?token=randomtoken123; contact=https://example.com/support)
```
Comment (JonLundy) [inline docs/_posts/2021-01-06-useragentextension.md:56]: A way we can firm up the URL perhaps we could use the sha256 of the target twtxt.txt URL

so if following `https://txt.sour.is/user/xuu/twtxt.txt` the request could be `https://example.com/whoFollows/6e62fb7419d9eebf3c59dc08d6c6e07b38c6cba5f0046df28239c847652a4423`

Comment (der-lyse) [reply docs/_posts/2021-01-06-useragentextension.md:56]: Hm, I'm not sure if I understand you correctly, but if we define to construct the Who Follows URL by simply hashing the feed URL, we're running exactly into the concerns in the "Security Considerations" section. Anybody could just find out about a certain feed's followers on a specific web client. The attackers would just have to also hash the target feed URL they're interested in to get the token with which they can query all the subscribers. To me this is a privacy violation.

I'm probably missing something here, because your hash doesn't line up: :-)
```bash
echo https://txt.sour.is/user/xuu/twtxt.txt | sha256sum
0ca0251c49492aafc9fefe69674d5261bcddbe18a205872a716695cc4b099e7c  -
```

Comment (JonLundy) [reply docs/_posts/2021-01-06-useragentextension.md:56]: > I'm probably missing something here, because your hash doesn't line up: :-)
> 
> ```shell
> echo https://txt.sour.is/user/xuu/twtxt.txt | sha256sum
> 0ca0251c49492aafc9fefe69674d5261bcddbe18a205872a716695cc4b099e7c  -
> ```

you are adding a new line. try `echo -n "https://txt.sour.is/xuu.txt" | sha256sum`

**Comment by JonLundy** [comment]
> Anybody could just find out about a certain feed's followers on a specific web client.

ah..i see.

File: docs/_posts/2021-01-06-useragentextension.md (hunk 11 above)
Comment (prologic) [reply docs/_posts/2021-01-06-useragentextension.md]: Let's go with the proposed:

```
User-Agent: twtxt/0.1.0 (~https://example.com/whoFollows?token=randomtoken123; contact=support@example.com)
```

File: docs/_posts/2021-01-06-useragentextension.md (hunk 12 above)
Comment (prologic) [reply docs/_posts/2021-01-06-useragentextension.md:56]: Yeah this was originally designed to be a fairly weak form of "token based auth", where it would thwart and prevent brute force attacks. I'm not sure your idea @JonLundy would work here?

Comment (der-lyse) [reply docs/_posts/2021-01-06-useragentextension.md:56]: > you are adding a new line. try `echo -n "https://txt.sour.is/xuu.txt" | sha256sum`

Indeed, and the path of the URL is different as well. ;-)

**Comment by der-lyse** [comment]
After a quick chat with @prologic I changed the syntax to make it more obvious, that the contact information is some kind of URI. Personally, I'd rather use a URL over an e-mail address if I would operate a multi user client because of spam reasons.

Another thing regarding spam towards twtxt.net rather than this spec: We could add another setting for an e-mail address which is specifically used just in this UA header. It could default to the admin e-mail setting, this way the addresses could be different for different purposes and private/public uses.

**Review by prologic — DISMISSED** [review]

**Comment by der-lyse** [comment]
Okay friends, I just implemented the specification, please have a look! I'm not a Go programmer, so I'm not sure whether the `CachedItem.Value` type change is okay or evil. I wanted to avoid a yummy copy pasta.

File: internal/whofollows_handler.go [hunk 13]
```diff
@@ -67,15 +59,27 @@ func (s *Server) WhoFollowsHandler() httprouter.Handle {
 			return
 		}
 
+		nick := ""
 		for _, user := range users {
 			if !user.IsFollowersPubliclyVisible && !ctx.User.Is(user.URL) {
```
Comment (der-lyse) [inline internal/whofollows_handler.go:64]: I'm not sure what this `ctx.User.Is(user.URL)` is all about, I believe it can be dropped, but I actually don't have a clue.

File: internal/whofollows_handler.go [hunk 14]
```diff
@@ -67,15 +59,27 @@ func (s *Server) WhoFollowsHandler() httprouter.Handle {
...
 			if user.Follows(uri) {
 				followers[user.Username] = user.URL
+				if nick == "" {
+					for n, url := range user.Following {
+						if url == uri {
+							nick = n
+							break
+						}
+					}
+				}
```
Comment (der-lyse) [inline internal/whofollows_handler.go:77]: Not sure whether there's a better way to determine the feed's nick.

File: internal/ttlcache.go [hunk 15]
```diff
@@ -5,22 +5,22 @@ import (
 	"time"
 )
 
-type CachedItem struct {
-	Value  int
-	Expiry time.Time
+type cachedItem struct {
+	value  interface{}
```
Comment (der-lyse) [inline internal/ttlcache.go:9]: Is the change from `int` to `interface{}` alright? I heard @JonLundy cry in IRC about it. ;-) I want to avoid duplicating the `cacheItem` and `TTLCache`, just to have the opportunity to use an `int` and a `string` value. The downside of my approach is obviously, that programmers can now try to obtain an int as string and vice versa, which compiles fine and just returns the default `0` or `""`, but the compiler doesn't catch an obvious misuse.

```go
ttlcache := NewTTLCache(5 * time.Minute)
ttlcache.Set("foo", 42)
ttlcache.GetString("foo") // → ""
ttlcache.Get("foo")       // → 42

ttlcache.SetString("bar", "baz")
ttlcache.Get("bar")       // → 0
ttlcache.GetString("bar") // → "baz"

File: internal/utils.go [hunk 16]
```diff
@@ -1409,13 +1409,15 @@ func URLForTask(baseURL, uuid string) string {
 	)
 }
 
-func URLForWhoFollows(baseURL string, feed types.Feed) string {
-	token := GenerateToken()
-
+func URLForWhoFollows(baseURL string, feed types.Feed, feedFollowers int) string {
 	return fmt.Sprintf(
-		"%s/whoFollows?uri=%s&nick=%s&token=%s",
+		"%s/whoFollows?followers=%d&token=%s",
```
Comment (der-lyse) [inline internal/utils.go:1414]: I considered to shorten the query parameter names to "n" and "t" to save even more space, but then decided to keep them readable for the server operators having a look at their access logs.

File: internal/whofollows_handler.go (hunk 13 above)
Comment (prologic) [reply internal/whofollows_handler.go:64]: I _think_ this should be kept, it allows you to view the data if you are hitting it yourself, rather than anonymously or another use.

File: internal/whofollows_handler.go (hunk 14 above)
Comment (prologic) [reply internal/whofollows_handler.go:77]: There's a reverse mapping called `User.remotes` Try that. It mapes from `url -> nick`

File: internal/ttlcache.go (hunk 15 above)
Comment (prologic) [reply internal/ttlcache.go:9]: This this is fine.

File: internal/whofollows_handler.go (hunk 14 above)
Comment (der-lyse) [reply internal/whofollows_handler.go:77]: Thanks! `User.sources` works perfectly. I changed it in the last commit.

File: internal/whofollows_handler.go (hunk 13 above)
Comment (der-lyse) [reply internal/whofollows_handler.go:64]: Well, as explained on IRC, for this super strange edge case where a logged in user grabbed the URL from his logs to test this, we could keep this check. :-)

**Comment by prologic** [comment]
Testing this locally with a local nginx container with a couple of files that simulate an external feed:

```#!console
~/Public
 0
$ ls -lah
total 256K
drwxr-xr-x+   5 prologic staff  160 Feb  6 10:14 .
drwxr-xr-x+ 107 prologic staff 3.4K Feb  6 10:12 ..
-rw-r--r--    1 prologic staff    0 Dec 23  2019 .localized
-rw-r--r--    1 prologic staff 252K Feb  6 10:14 avatar.png
-rw-r--r--    1 prologic staff   39 Feb  6 10:12 twtxt.txt
```

**Comment by prologic** [comment]
Running this simply as:

```#!console
$ docker run -i -t --rm -v "$PWD":/usr/share/nginx/html:ro -p 9000:80 nginx
```

using Docker for Mac.

And following:

- Nick: test
- URL: http://localhost:9000/twtxt.txt

On a local dev pod.

**Comment by prologic** [comment]
Yup this works; Merging:

```#!console
$ curl -q -s -H 'Accept: application/json' -o - 'http://10.0.0.109:8000/whoFollows?followers=2&token=AlePVx8ZP' | jq '.'
{
  "admin": "http://10.0.0.109:8000/user/admin/twtxt.txt",
  "kate": "http://10.0.0.109:8000/user/kate/twtxt.txt"
}
```

**Review by prologic — APPROVED** [review]

**Comment by prologic** [comment]
Merging anyway despite the GoReleaser build failures (_will fix later_)
